### PR TITLE
Scaled up from 1

### PR DESCRIPTION
**Pull request by wahyuoi**
Signed-off-by: Gede Wahyu <tokekbesi@gmail.com>

## Description
- Fix replica calculation to enable scaled up from 1 when max replica or scale factor is small

## Motivation and Context
- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

When I have this config:
- current replica = 1
- max replica = 5
- scale factor = 30

it will never scaled up from 1. 

It will scaled up iff #MaxReplica * #ScaleFactor >= 200

## How Has This Been Tested?
I create new test case and already run test for `alerthandler_test.go`
Also tested on kubernetes using hey tool and figlet with max-replicas=5 and scale-factor=1
Result :
```
2018/12/01 18:02:13 [Scale] function=figlet 1 => 2.
2018/12/01 18:02:53 [Scale] function=figlet 2 => 3.
2018/12/01 18:03:33 [Scale] function=figlet 3 => 4.
2018/12/01 18:04:13 [Scale] function=figlet 4 => 5.
```
Result using default configuration, max-replicas=20 and scale-factor=20:
```
2018/12/01 18:21:28 [Scale] function=figlet 1 => 5.
2018/12/01 18:22:08 [Scale] function=figlet 5 => 9.
2018/12/01 18:22:48 [Scale] function=figlet 9 => 13.
2018/12/01 18:23:28 [Scale] function=figlet 13 => 17.
2018/12/01 18:24:08 [Scale] function=figlet 17 => 20.
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Templum** [comment]
Looks good to me. That should address your issue.

**Comment by wahyuoi** [comment]
My bad, didn't use latest code from upstream and introduce single file conflict.
Already fixed, thanks

File: gateway/handlers/alerthandler.go [hunk 1]
```diff
@@ -100,14 +100,10 @@ func CalculateReplicas(status string, currentReplicas uint64, maxReplicas uint64
 	step := uint64((float64(maxReplicas) / 100) * float64(scalingFactor))
 
 	if status == "firing" && step > 0 {
-		if currentReplicas == 1 {
-			newReplicas = step
```
Comment (ivanayov) [inline gateway/handlers/alerthandler.go]: The issue here is that `step` in rounded down (1.5 -> 1). I wonder if it's better to change the logic, or fix it by rounding up before casting.

Comment (wahyuoi) [reply gateway/handlers/alerthandler.go]: Great, just found 1 case that makes it fail.
When maxReplica*scaleFactor < 100. 

Just fixed, thanks @ivanayov

Comment (wahyuoi) [reply gateway/handlers/alerthandler.go]: If we fix by rounding up only, it still fail on new case I provide.
IMO, we still need to change the logic on line 103.

Comment (alexellis) [reply gateway/handlers/alerthandler.go]: ping @ivanayov

**Comment by alexellis** [comment]
@ivanayov please could you sign-off on this or let us know if we need some additional changes / test-cases?

File: gateway/handlers/alerthandler.go (hunk 1 above)
Comment (ivanayov) [reply gateway/handlers/alerthandler.go]: I agree. The old logic is wrong for one more thin: we ignore `currentReplicas` when it's 1 and it's not proper to say `step` in that case

**Comment by wahyuoi** [comment]
@alexellis PR message already updated, there is additional end-to-end test result using latest-dev gateway.

**Comment by alexellis** [comment]
What about the default case of 1 min 20 max and scale factor of 20?

File: gateway/handlers/alerthandler_test.go [hunk 2]
```diff
@@ -99,3 +99,25 @@ func TestBackingOff(t *testing.T) {
 		t.Fail()
 	}
 }
+
+func TestScaledUpFrom1(t *testing.T) {
+	currentReplicas := uint64(1)
+	maxReplicas := uint64(5)
+	scalingFactor := uint64(30)
+	newReplicas := CalculateReplicas("firing", currentReplicas, maxReplicas, scaling.DefaultMinReplicas, scalingFactor)
+	if newReplicas == currentReplicas {
```
Comment (rgee0) [inline gateway/handlers/alerthandler_test.go]: This test will pass if current > new

File: gateway/handlers/alerthandler_test.go [hunk 3]
```diff
@@ -99,3 +99,25 @@ func TestBackingOff(t *testing.T) {
...
+		t.Fail()
+	}
+}
+
+func TestScaledUpWithSmallParam(t *testing.T) {
+	currentReplicas := uint64(1)
+	maxReplicas := uint64(4)
+	scalingFactor := uint64(1)
+	newReplicas := CalculateReplicas("firing", currentReplicas, maxReplicas, scaling.DefaultMinReplicas, scalingFactor)
+	if newReplicas == currentReplicas {
```
Comment (rgee0) [inline gateway/handlers/alerthandler_test.go]: This test will pass if current > new

**Comment by LucasRoesler** [comment]
This looks good to me, are we ready to merge it?

**Comment by alexellis** [comment]
Can we get a final check from @rgee0 first please? He's been giving feedback and working with the contributor.

**Comment by alexellis** [comment]
Derek set milestone: 0.9.12

**Comment by alexellis** [comment]
@rgee0 @ivanayov please can you confirm if are you available to test this end to end when released under 0.9.12?

**Pull request merged**